### ISSUE TITLE
Generic Lobby

### ISSUE THREAD

**Issue by nicolodavis** [nicolodavis/boardgame.io#354]
Ideas for improvement:

[ ] Poll for changes to state so that refreshes are not required.
[ ] Option to automatically remove room if all users leave the room
[ ] Add optional unique name check
[ ] Add optional limit for rooms (will disable 'create room' button if rooms number reached limit)
[ ] Provide ability to customize ui when waiting for an API call (i.e. loading state)
[ ] Rename `gameInstances` to `rooms` (easier to understand).

**Comment by flamecoals** [comment]
Why don't we use websockets instead of polling?

**Comment by blunket** [comment]
IMO polling is better for this, sockets are overkill when it's just a lobby and doesn't need to be real time

**Comment by lehaSVV2009** [comment]
btw, I also thought that lobby API might be improved a bit as well. To smth like that:
`GET /games
GET /games/my-game/rooms
POST /games/my-game/rooms
POST /games/my-game/rooms/123/join
POST /games/my-game/rooms/123/leave
DELETE /games/my-game/rooms/123`

**Comment by flamecoals** [comment]
One use case I want to bring to everybody's attention... My project will have several games, and I want people to choose the game after creating the lobby. And after a game is done, I want people to be able to play another game with the same people.

Is it ok if we key only by lobby id instead of both game and id? I can work on this latter

**Comment by blunket** [comment]
@flamecoals What do you mean? Multiple lobbies? Couldn't you just filter which games to display however you like? Or do you mean the games will be hosted on many different game servers too, or something like that?

Right now players choose the game to play IN the lobby, they don't create lobbies.

**Comment by nicolodavis** [comment]
Yes, I'm envisioning just one global lobby, and you can create games inside that lobby (each game can be a different kind). Players then join the game you created in order to play.

**Comment by flamecoals** [comment]
@blunket I meant one global lobby like Nicolo said. Basically a "room" persisting across matches too, of different types of games.

Use case:
- We and my friend Joe want to play some games, but we dont know exactly what
- We create a lobby for us, and we agree in playing Chess inside the lobby
- After the chess match is done, we go back to the same lobby and agree in playing tic tac toe

and so on

**Comment by blunket** [comment]
I see, I think you mean sort of like a party system? That'd be cool.

**Comment by flamecoals** [comment]
Hey folks, I started a document for documenting and designing all these ideas we were talking about:
https://docs.google.com/document/d/1K0xDlyS_80WJic7GfNobl-SujfC8_LiB55-JJnAawNM/edit?usp=drivesdk

Still WIP, but feel free to comment or add anything I am forgetting. I will update this thread whenever I think it is ready for review.

**Comment by flamecoals** [comment]
Hey folks, I finished the first pass of the design document for the lobby:
https://docs.google.com/document/d/1K0xDlyS_80WJic7GfNobl-SujfC8_LiB55-JJnAawNM/edit#

Just so nobody freaks out, I won''t do this in a single PR. I just want to get an agreement that this is the desired state and we can do small PRs towards that goal. See Development Plan section.

@nicolodavis It would be great if you could give me some clarity around "Open questions" section

**Comment by nicolodavis** [comment]
@flamecoals Sent you some comments on your doc.

**Comment by flamecoals** [comment]
Ok, @nicolodavis, I replied to them. Given the points you asked are not questioning the overall idea, I think we are more or less in agreement on the path forward. I will start sending PRs so we can start implementing this. :)

**Comment by nicolodavis** [comment]
I think we should have agreement on the user data storage part first, because that's a critical design feature.

Feel free to implement any of the items that don't touch that yet though.

**Comment by flamecoals** [comment]
Ok :)

**Comment by francoijs** [comment]
Bravo for the lobby design proposal and the documentation!
I've added a few things in the description of the current behavior.

**Comment by flamecoals** [comment]
@francoijs Thanks for the help! Approved all your suggestions. I will probably do the first PR soon on this.

**Comment by nicolodavis** [comment]
@flamecoals @francoijs @jasonharrison 

Do you guys have thoughts about what to do about the Lobby in the Svelte world (after #432)?  We can migrate the lobby rendering components to Svelte (easy), but I'm not sure how to then hook it into the board components that users create, which might be in React / Angular or whatever.

We want the lobby to be in something generic like Svelte so that folks using the vanilla JS client also have access to it.  However, if the user chooses to code their board in React, we want the lobby to be able to instantiate it as well.

**Comment by nicolodavis** [comment]
Perhaps we should use a generic `render` callback of some sort that the user can use to the render the game using any technology of their choice?

#### Current State

We pass in the React components to the lobby directly (which is itself a React component).  The Lobby then knows which component to render based on the game selected.

```js
<Lobby gameComponents={ ... } />
```

#### Proposal

Once we migrate the Lobby to Svelte (or even if we keep it in React for that matter), I think we should expose it as a function to the end user (rather than a component).

```js
import { Lobby } from 'boardgame.io/lobby';

const lobby = Lobby({
  games: [ ... ],

  render: (gameName, playerID, gameID, root) => {
    // This function is called once the player has chosen the game and hits "Play".
    // We can render a component using any framework of our choice
    // (including React as shown below).  We should also consult gameName
    // before deciding which component to use (not shown in this example).
    // root is just a HTMLElement on which we render our client.
    ReactDOM.render(<TicTacToe gameID={gameID} playerID={playerID} />, root);
  },
});

lobby.run();
```

We can also preserve the convenience of our current approach by providing a bundled renderer for React:

```js
const lobby = Lobby({
  games: [ ... ],

  render: ReactRenderer({ gameComponents: { ... } }),
});
```

**Comment by delucis** [comment]
Is [`src/lobby/connection.js`](https://github.com/nicolodavis/boardgame.io/blob/master/src/lobby/connection.js) currently useable as a vanilla “lobby client”? Mirroring trying to develop the vanilla game client (supporting events etc.), it would be nice if there was something users could plug their own lobby rendering into that provides the basic API (list, join, leave, etc.)

**Comment by delucis** [comment]
Also, looking at [this](https://github.com/jesseskinner/svelte-react), it looks like an extra dependency might be needed to allow a Svelte Lobby to handle React components (and more dependencies for more frameworks). Maybe it would be better to have framework-specific packages that sit on top of the vanilla client and lobby APIs and provide components in React, Vue, etc.? (Svelte Client and Lobby components could still be used by those not using frameworks.)

**Comment by nicolodavis** [comment]
Why would we need to pull in `react-svelte`?  Wouldn't the React dependencies be the responsibility of the developer using boardgame.io?

```js
import { Lobby } from 'boardgame.io/lobby';

// These are in client code.
import React from 'react';
import ReactDOM from 'react-dom';

const lobby = Lobby({
  games: [ ... ],

  render: (gameName, playerID, gameID, root) => {
    ReactDOM.render(<TicTacToe gameID={gameID} playerID={playerID} />, root);
  },
});

lobby.run();
```

**Comment by delucis** [comment]
Ohhhh, I see — `react-svelte` is if you need to use React in your Svelte app before compiling it. Then the above looks great!

Do you think there should be some documentation on how to use `LobbyConnection` directly for those who don’t want to use the provided component?

**Comment by nicolodavis** [comment]
`LobbyConnection` is just a simple REST client that invokes our API.  Do you think there is much value in making that externally visible?  I would think that people would prefer to use their favorite library to call our REST API directly as opposed to using our client wrapper.

**Comment by nicolodavis** [comment]
My thinking was that we just provide the REST API (which can be consumed in many ways).  In addition to that, we also provide a default lobby implementation (that can be styled).  If someone wants to build a completely different lobby, they can go ahead and build that on top of our REST API.

Maybe what we should be focusing on is making our existing lobby generic enough that people won't see the need to re-implement it?
